pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -506,13 +506,23 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
 		return nil
 	}
 
-	if pipelineState.IsDone() && pr.IsDone() {
+	if pipelineState.IsDone() && pipelineState.IsFinalTasksDone() && pr.IsDone() {
 		c.timeoutHandler.Release(pr)
 		c.Recorder.Event(pr, corev1.EventTypeNormal, eventReasonSucceeded, "PipelineRun completed successfully.")
 		return nil
 	}
 
-	for _, rprt := range pipelineState {
+	pipelineTasks := resources.PipelineRunState{}
+	finalTasks := resources.PipelineRunState{}
+	for _, task := range pipelineState {
+		if task.PipelineTask != nil {
+			pipelineTasks = append(pipelineTasks, task)
+		} else if task.FinalPipelineTask != nil {
+			finalTasks = append(finalTasks, task)
```

Comment:
would it make sense for PipelineRunState to hold both final and regular tasks so you dont need to use 2 different objects, and you dont need to iterate over them separately?

Reply:
> This is looking great! I think we're all almost totally agreed about the feature itself so most of my comments are about the implementation.

🎉 🎉 🎉 
 
> Main themes:
> 
> * I'm not sure about allowing "from" in finally tasks, feels a bit confusing to me when we don't allow results (e.g. imagine a git resource that is "from" a task that builds - if the build task executes, the data on disc for the resource may be different than if it doesnt)

Yup, I agree, task results and `from` fall under the same category and can be added incrementally. 

> * I think its worth trying to avoid duplicating the validation logic as much as possible, since it leads to some nasty bugs, for example we ran into that with results (#2471) - we're still duplicating some logic there so it might not be completely avoidable, but if we can avoid it its worth the effort

yup, on it, have commented below as well, how about something like [this](https://github.com/tektoncd/pipeline/compare/master...pritidesai:refactor-pipeline-validation), any ideas or feedback is appreciated, I am still learning `v1alpha1` and `v1beta1` APIs.

> * Better to add functions in separate packages than add more functions to the reconciler

Sure, still learning about reconcilers 🤪

> * If we get the structure of PipelineRunState right, i bet we can avoid some of the duplicated code in the reconciler
> * Is it possible for the logic that determines what runs next to become aware of finally? If the logic that returns the next tasks to run can handle Finally tasks, i bet we need a lot less special casing for finally

Yup, see my comment [below](https://github.com/tektoncd/pipeline/pull/2437#discussion_r420534851), I am working on proposal which can be part of separate PR and is applicable without finally as well.

> It looks like the coverage generally went down, and we want it to generally go up or stay the same so it might be worth digging into that a bit once we're more sure about the details of the implementation:
> 
> ![image](https://user-images.githubusercontent.com/432502/81101914-ce1a4980-8edc-11ea-8067-e5ec0249d561.png)

yup, not happy at all, I have lot more unit testing which is not part of this PR, was nervous with the sizing and growing number of LOC. Splitting is the key, will have separate PRs.

Reply:
have separated PRs and added more unit tests to make sure coverage doesnt go down 😄